pypdf/_page_labels.py

```diff
@@ -132,49 +168,21 @@ def index2label(reader: PdfCommonDocProtocol, index: int) -> str:
         return str(index + 1)  # Fallback
     number_tree = cast(DictionaryObject, root["/PageLabels"].get_object())
     if "/Nums" in number_tree:
-        # [Nums] shall be an array of the form
-        #   [ key 1 value 1 key 2 value 2 ... key n value n ]
-        # where each key_i is an integer and the corresponding
-        # value_i shall be the object associated with that key.
-        # The keys shall be sorted in numerical order,
-        # analogously to the arrangement of keys in a name tree
-        # as described in 7.9.6, "Name Trees."
-        nums = cast(ArrayObject, number_tree["/Nums"])
-        i = 0
-        value = None
-        start_index = 0
-        while i < len(nums):
-            start_index = nums[i]
-            value = nums[i + 1].get_object()
-            if i + 2 == len(nums):
-                break
-            if nums[i + 2] > index:
-                break
-            i += 2
-        m = {
-            None: lambda n: "",
-            "/D": lambda n: str(n),
-            "/R": number2uppercase_roman_numeral,
-            "/r": number2lowercase_roman_numeral,
-            "/A": number2uppercase_letter,
-            "/a": number2lowercase_letter,
-        }
-        # if /Nums array is not following the specification or if /Nums is empty
-        if not isinstance(value, dict):
-            return str(index + 1)  # Fallback
-        start = value.get("/St", 1)
-        prefix = value.get("/P", "")
-        return prefix + m[value.get("/S")](index - start_index + start)
-    if "/Kids" in number_tree or "/Limits" in number_tree:
-        logger_warning(
-            (
-                "/Kids or /Limits found in PageLabels. "
-                "This is not yet supported."
-            ),
-            __name__,
-        )
-    # TODO: Implement /Kids and /Limits for number tree
-    return str(index + 1)  # Fallback if /Nums is not in the number_tree
+        return get_label_from_nums(number_tree, index)
+    if "/Kids" in number_tree and not isinstance(number_tree["/Kids"], NullObject):
+        # number_tree = {'/Kids': [IndirectObject(7333, 0, 140132998195856), ...]}
+        kids: List[DictionaryObject] = cast(List[DictionaryObject], number_tree["/Kids"])
+        for kid in kids:
+            # kid = {'/Limits': [0, 63], '/Nums': [0, {'/P': 'C1'}, ...]}
+            limits: List[int] = cast(List[int], kid["/Limits"])
+            if limits[0] <= index <= limits[1]:
+                return get_label_from_nums(kid, index)
```

Comment:
We could have multiple levels when merging multiple documents for example.  It might be good to consider a recursive approach.

Reply:
Could you please elaborate? Running something like

```python3
writer = PdfWriter(clone_from=reader)
writer.append_pages_from_reader(reader)
print(writer.page_labels)
```

will not generate any labels for the second set of pages and I have no other files to test. As long as we do not have an actual example, this is hard to grasp and verify the correct behavior. I would argue that it is preferable to leave this out for now and reconsider this iff somebody reports an actual example.

Reply:
this should be good:
[tt1.pdf](https://github.com/py-pdf/pypdf/files/14842446/tt1.pdf)

Reply:
Given that both Evince and PDFBox Debugger support this, I have just pushed a change. There is still some difference due to the incomplete limits, but this is an implementation detail regarding input data issues.